library/core/src/hint.rs

```diff
@@ -322,6 +322,11 @@ pub fn spin_loop() {
 ///
 /// </div>
 ///
+/// You should not be able to write a better `black_box` than the standard library provides. If you
+/// can write a function that looks like `black_box` but is more effective at hiding the value of
+/// the argument from optimizations, that would be a bug and should be reported. It just would not
+/// be a *security* bug.
```

Comment:
I'm not sure if that's quite true. For example user code might choose to use FFI and setup linking to evade LTO, I don't think std is going to jump through those kinds of hoops since it can't control linking of generic instantiations.

Reply:
Perhaps "you should not be able to use other parts of the standard library to write a better..." then?

Reply:
I'm not sure what you mean by evading LTO. If LTO can see through black_box it's already quite non-functional for benchmarking.

Reply:
A weak blackbox implementation might prevent DCE of the input but not optimizations based on properties of the output value.

You can also replace LTO with other post-link optimizations such as wasm JIT and user code taking steps avoid that.

Reply:
A JIT can break any black box so I'm not sure what point you are trying to make.

Reply:
A JIT will have blind spots or even explicit escape hatches and you can make a targeted blackbox using those.

The point I'm making is that users may be "able to write a better `black_box`" under specific circumstances and the standard library won't adopt those.

Reply:
> A weak blackbox implementation might prevent DCE of the input but not optimizations based on properties of the output value.

that's an example where you could do better, and it would indeed be a buggy implementation that should be reported, as the docs say.

Reply:
No, I meant that as an example for a hypothetical backend or target where that's the best we can do.
E.g. by making a value "escape" to a static we could prevent DCE but the backend lacks a way to make the value opaque because it's a single-threaded closed-world environment that has no concept of environment other than some limited IO so even volatile would get lowered to plain stores and loads.